### PR TITLE
Add 90overlay-root module on request of Debian Maintainer

### PR DESCRIPTION
**Pull request by danimo**
Putting this here for discussion.

### CODE REVIEW

File: modules.d/90overlay-root/overlay-mount.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
+
+# make a read-only nfsroot writeable by using overlayfs
+# the nfsroot is already mounted to $NEWROOT
+# add the parameter rootovl to the kernel, to activate this feature
+
+. /lib/dracut-lib.sh
+
+if ! getargbool 0 rootovl ; then
```
Comment (danimo) [inline modules.d/90overlay-root/overlay-mount.sh:9]: This should probably become `rd.rootovl` (and keep `rootovl` as a compat parameter). @haraldh?

**Comment by danimo** [comment]
//cc @Vogtinator This may also be of interest to you

File: modules.d/90kernel-modules-pinctrl/module-setup.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (Vogtinator) [inline modules.d/90kernel-modules-pinctrl/module-setup.sh:1]: Unrelated?

File: modules.d/90overlay-root/overlay-mount.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return
+fi
+
+modprobe overlay
+
+# a little bit tuning
+mount -o remount,nolock,noatime $NEWROOT
+
+# Move root
```
Comment (Vogtinator) [inline modules.d/90overlay-root/overlay-mount.sh:18]: `mount --make-private /` would fix this. Though I don't know whether that breaks something else.

Comment (Vogtinator) [reply modules.d/90overlay-root/overlay-mount.sh:18]: If making `/` private is not an option, please move the pseudo-move into a function

File: modules.d/90overlay-root/overlay-mount.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
+
+# make a read-only nfsroot writeable by using overlayfs
+# the nfsroot is already mounted to $NEWROOT
+# add the parameter rootovl to the kernel, to activate this feature
+
+. /lib/dracut-lib.sh
+
```
Comment (Vogtinator) [inline modules.d/90overlay-root/overlay-mount.sh:8]: I recommend `set -eu`

File: modules.d/90overlay-root/overlay-mount.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# Move root
+# --move does not always work. Google >mount move "wrong fs"< for
+#     details
+mkdir -p /live/image
+mount --bind $NEWROOT /live/image
+umount $NEWROOT
+
+# Create tmpfs
+mkdir /cow
+mount -n -t tmpfs -o mode=0755 tmpfs /cow
```
Comment (Vogtinator) [inline modules.d/90overlay-root/overlay-mount.sh:27]: Why `-n` here?

File: modules.d/90overlay-root/overlay-mount.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+mkdir /cow/work /cow/rw
+
+# Merge both to new Filesystem
+mount -t overlay -o noatime,lowerdir=/live/image,upperdir=/cow/rw,workdir=/cow/work,default_permissions overlay $NEWROOT
```
Comment (Vogtinator) [inline modules.d/90overlay-root/overlay-mount.sh:31]: Why `noatime`?

**Review by Vogtinator — CHANGES_REQUESTED** [review]

**Comment by Vogtinator** [comment]
I've used something like this for years several years ago, but could not use overlayfs on NFSv4 due to ACLs in xattrs. Is that fixed meanwhile or does this only work with either NFSv3 or NFSv4 without ACLs?

File: modules.d/90kernel-modules-pinctrl/module-setup.sh (hunk 2 above)
Comment (danimo) [reply modules.d/90kernel-modules-pinctrl/module-setup.sh:1]: Indeed. my fault.

File: modules.d/90overlay-root/module-setup.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+check() {
+    # do not add modules if the kernel does not have overlayfs support
+    [ -d $dracutsysrootdir/lib/modules/$kernel/kernel/fs/overlayfs ] || return 1
```
[the remark this answers is not in the document]
Comment (danimo) [reply modules.d/90overlay-root/module-setup.sh:5]: I was also wondering about this. What's a good way to check here? Purely checking for functionality might not work, because it might not be loaded.

[the remark this answers is not in the document]
Comment (Vogtinator) [reply modules.d/90overlay-root/module-setup.sh:5]: Just the `instmods` in `installkernel` should be enough. It's impossible to guarantee 100% that if `check` succeeds that it actually works anyway.

File: modules.d/90overlay-root/module-setup.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+depends() {
+    # We do not depend on any modules - just some root
+    return 0
+}
+
+# called by dracut
+installkernel() {
+    instmods overlay
```
Comment (Vogtinator) [inline modules.d/90overlay-root/module-setup.sh:15]: Might need `hostonly=''`

File: modules.d/90overlay-root/module-setup.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (danimo) [reply modules.d/90overlay-root/module-setup.sh:5]: Good point.

File: modules.d/90overlay-root/module-setup.sh (hunk 8 above)
Comment (Mrfai) [reply modules.d/90overlay-root/module-setup.sh:15]: You  are right, overlayfs and NFSv4 is a little bit tricky.

As of 2019 overlayfs  can only work with NFS v4 if acl(5) support is disabled on the filessystem on the NFS server side. This can be done by remounting the nfsroot on the NFS server with noacl. I'm not sure if newer kernel version still need this.

Comment (danimo) [reply modules.d/90overlay-root/module-setup.sh]: Included. Keeping open because of the unrelated comment.

File: modules.d/90overlay-root/module-setup.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (danimo) [reply modules.d/90overlay-root/module-setup.sh]: Removed.

File: modules.d/90overlay-root/overlay-mount.sh (hunk 5 above)
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:27]: IIRC, at  this boot time /etc/mtab is not writeable.

Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:27]: I my usecase, I always mount a read-only nfsroot.

File: modules.d/90overlay-root/overlay-mount.sh (hunk 6 above)
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:31]: noatime is just a small optimisation. You are remove this.

File: modules.d/90overlay-root/overlay-mount.sh (hunk 5 above)
Comment (Vogtinator) [reply modules.d/90overlay-root/overlay-mount.sh:27]: AFAIK / is the initrd here, so tmpfs. It's post likely linked to `/proc/self/mounts` anyway and I'd expect that the `mount --bind` calls also write to mtab.

Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:27]: I'll try to check if it works without -n.

Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:27]: -n can be removed, also noatime is not needed

File: modules.d/90overlay-root/overlay-mount.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# the nfsroot is already mounted to $NEWROOT
+# add the parameter rootovl to the kernel, to activate this feature
+
+. /lib/dracut-lib.sh
+
+if ! getargbool 0 rootovl ; then
+    return
+fi
+
+modprobe overlay
```
[the remark this answers is not in the document]
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:13]: modprobe is not needed. I just did a test.

File: modules.d/90overlay-root/overlay-mount.sh (hunk 4 above)
Comment (Vogtinator) [reply modules.d/90overlay-root/overlay-mount.sh:8]: Unfortunately, yes... IMO it's still better to bail out early if something fails than to continue like nothing bad happened. In the case where the mount fails, it would actually exit 1 (by chance?) because the last `mount --move` would fail and such set the script's return value. The errors until that point would be completely ignored though.

File: modules.d/90overlay-root/overlay-mount.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+mkdir -p /live/image
+mount --move $NEWROOT /live/image
+
+# Create tmpfs
+mkdir /cow
+mount -t tmpfs -o mode=0755 tmpfs /cow
+mkdir /cow/work /cow/rw
+
+# Merge both to new Filesystem
+mount -t overlay -o lowerdir=/live/image,upperdir=/cow/rw,workdir=/cow/work,default_permissions overlay $NEWROOT
```
Comment (Vogtinator) [inline modules.d/90overlay-root/overlay-mount.sh:23]: It might be a performance (and memory) improvement to use `redirect_dir=on,metacopy=on` here to avoid copying of data to the tmpfs. Those options are used from the kernel configuration if unspecified, so depending on the setup that might be the default.

File: modules.d/90overlay-root/overlay-mount.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+# Move root
+# --move does not always work. Google >mount move "wrong fs"< for
+#     details
+mkdir -p /live/image
+mount --bind $NEWROOT /live/image
+umount $NEWROOT
+
+# Create tmpfs
+mkdir /cow
```
[the remark this answers is not in the document]
Comment (Vogtinator) [reply modules.d/90overlay-root/overlay-mount.sh:18]: `/sysroot-overlay` maybe?

[the remark this answers is not in the document]
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:18]: Good idea. But I leave this to the RH people what to choose.

File: modules.d/90overlay-root/overlay-mount.sh (hunk 4 above)
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh:8]: First test using set -eu:
lib/dracut-lib.sh: line 61: DRACUT_SYSTEMD unbound variable
_newoption: unbound variable
DRACUT_SYSTEMD: unbound variable
then kernel panic
 :-(

**Comment by Mrfai** [comment]
> Great, I will also look at the README. -- regards Thomas

I forgot what we wanted to change in README. Any hints?

File: modules.d/90overlay-root/overlay-mount.sh (hunk 4 above)
Comment (Vogtinator) [reply modules.d/90overlay-root/overlay-mount.sh:8]: It may have to be done after sourcing `lib/dracut-lib.sh`...

**Comment by Vogtinator** [comment]
> > Great, I will also look at the README. -- regards Thomas
> 
> I forgot what we wanted to change in README. Any hints?

No idea, but I noticed that it doesn't mention that changes end up in tmpfs and are volatile.

**Review by haraldh — CHANGES_REQUESTED** [review]
Needs some documentation in `dracut.cmdline.7.asc` about the command line parameter mentioning the specific dracut module

**Comment by haraldh** [comment]
nobody?

**Comment by Mrfai** [comment]
This is my proposal for the dracut.cmdline.7.asc documentation

** rootovl**::
    Make a read-only rootfs writeable by adding a tmpfs on top via overlayfs.
    This is useful for read-only NFS root file systems.

**Comment by Mrfai** [comment]
I'm not sure about the workflow to update this PR.

**Comment by haraldh** [comment]
ah, it would require @danimo to make those changes, if @Mrfai file a new PR

**Comment by danimo** [comment]
Updated pull-request with all the feedback, shy of setting `+eu`. Graduate to full PR.

**Comment by danimo** [comment]
> Is the newline after mount -o in [f97baf7](https://github.com/dracutdevs/dracut/commit/f97baf7aea4074c10e00e8e62c199f7a70237d25) intentional?

Nope, that was a mistake. Fix.

**Comment by danimo** [comment]
> Needs some documentation in `dracut.cmdline.7.asc` about the command line parameter mentioning the specific dracut module

Can you switch to approval for this PR?

**Comment by Mrfai** [comment]
A minor change for README
/live/cow -> /live/sysroot-overlay

**Comment by danimo** [comment]
@haraldh @johannbg ready for merge

File: modules.d/90overlay-root/module-setup.sh [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+depends() {
+    # We do not depend on any modules - just some root
+    return 0
```
Comment (haraldh) [inline modules.d/90overlay-root/module-setup.sh:5]: Please make this `return 255`, so it is only included, if requested.

**Review by haraldh — CHANGES_REQUESTED** [review]

File: modules.d/90overlay-root/module-setup.sh (hunk 12 above)
Comment (Mrfai) [reply modules.d/90overlay-root/module-setup.sh]: I cannot find any other dracut module that returns 255 here, so I don't understand why this module should do it.

File: modules.d/90overlay-root/overlay-mount.sh [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+mkdir /sysroot-overlay/work /sysroot-overlay/rw
+
+# Merge both to new Filesystem
+mount -t overlay -o lowerdir=/live/image,upperdir=/sysroot-overlay/rw,workdir=/sysroot-overlay/work,redirect_dir=on,metacopy=on,default_permissions overlay "$NEWROOT"
+
+# Let filesystems survive pivot
+mkdir -p "$NEWROOT/live/sysroot-overlay"
+mkdir -p "$NEWROOT/live/image"
+mount --move /sysroot-overlay/rw "$NEWROOT/live/sysroot-overlay"
+mount --move /live/image "$NEWROOT/live/image"
```
[the remark this answers is not in the document]
Comment (Mrfai) [reply modules.d/90overlay-root/overlay-mount.sh]: Good idea. @danimo can you fix this?

File: modules.d/90overlay-root/module-setup.sh (hunk 12 above)
Comment (haraldh) [reply modules.d/90overlay-root/module-setup.sh]: Oops.. my error... I meant in `check()`

File: modules.d/90overlay-root/module-setup.sh [hunk 14]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+depends() {
```
Comment (haraldh) [inline modules.d/90overlay-root/module-setup.sh:3]: This should have:

```
check() {
    return 255
}
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: modules.d/90overlay-root/module-setup.sh (hunk 14 above)
Comment (haraldh) [reply modules.d/90overlay-root/module-setup.sh:3]: @danimo @Mrfai ... sorry, not in `depends()` but in `check()`

File: modules.d/90overlay-root/overlay-mount.sh [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+mount -t tmpfs -o mode=0755 tmpfs "$workdir"
+
+# Merge both to new Filesystem
+mount -t overlay -o "lowerdir=$lowerdir,upperdir=$upperdir,workdir=$workdir,redirect_dir=on,metacopy=on,default_permissions" overlay "$NEWROOT"
+
+# Let filesystems survive pivot
+mkdir -p "$NEWROOT$upperdir" "$NEWROOT$lowerdir" "$NEWROOT$workdir"
+mount --move "$upperdir" "$NEWROOT$upperdir"
+mount --move "$lowerdir" "$NEWROOT$lowerdir"
+mount --move "$workdir" "$NEWROOT$workdir"
```
Comment (haraldh) [inline modules.d/90overlay-root/overlay-mount.sh:33]: This does not work and should be removed.

**Review by haraldh — CHANGES_REQUESTED** [review]

File: modules.d/90overlay-root/overlay-mount.sh (hunk 15 above)
Comment (haraldh) [inline modules.d/90overlay-root/overlay-mount.sh:33]: This does not work and should be removed, because it is unnecessary with `/run`

**Review by haraldh — CHANGES_REQUESTED** [review]

File: modules.d/90overlay-root/overlay-mount.sh (hunk 15 above)
Comment (haraldh) [reply modules.d/90overlay-root/overlay-mount.sh:33]: `/run` will be moved by switch_root to `$NEWROOT/run` anyway

**Review by johannbg — APPROVED** [review]

File: man/dracut.cmdline.7.asc [hunk 16]
```diff
@@ -155,6 +155,10 @@ Misc
     started. If you have hardware, which takes a very long time to announce its
     drives, you might want to extend this value.
 
+**rd.root.overlay**::
```
Comment (LaszloGombos) [inline man/dracut.cmdline.7.asc:158]: It would be useful to provide guidance how this option is different from systemd.volatile=overlay and what are the tradeoffs ? One difference I noticed is that systemd.volatile=overlay would not leave the mounted filesystem visible. Is that the only difference ?

Comment (Mrfai) [reply man/dracut.cmdline.7.asc:158]: This module runs without systemd. It is used by the FAI project for several years. It seems that systemd.volatile only makes /etc and /var writeable, rd.root.overlay makes the whole / writeable.

Comment (LaszloGombos) [reply man/dracut.cmdline.7.asc:158]: Thanks for the clarification. I have not realized that systemd is an optional dependency for dracut.
